src/meta/src/controller/catalog.rs

```diff
@@ -0,0 +1,385 @@
+    /// ```sql
+    /// WITH RECURSIVE used_by_object_ids (used_by) AS
+    /// (
+    ///     SELECT used_by FROM object_dependency WHERE object_dependency.oid = $1
+    ///     UNION ALL
+    ///     (
+    ///         SELECT object_dependency.used_by
+    ///                     FROM object_dependency
+    ///             INNER JOIN used_by_object_ids
+    ///                     ON used_by_object_ids.used_by = oid
+    ///     )
+    /// )
+    /// SELECT DISTINCT used_by FROM used_by_object_ids ORDER BY used_by DESC;
+    /// ```
+    async fn list_used_by(&self, obj_id: i32) -> MetaResult<Vec<i32>> {
+        let inner = self.inner.read().await;
+
+        let cte_alias = Alias::new("used_by_object_ids");
+        let cte_return_alias = Alias::new("used_by");
+
+        let base_query = SelectStatement::new()
+            .column(object_dependency::Column::UsedBy)
+            .from(ObjectDependency)
+            .and_where(object_dependency::Column::Oid.eq(obj_id))
+            .to_owned();
+
+        let cte_referencing = Query::select()
+            .column((ObjectDependency, object_dependency::Column::UsedBy))
+            .from(ObjectDependency)
+            .join(
+                JoinType::InnerJoin,
+                cte_alias.clone(),
+                Expr::col((cte_alias.clone(), cte_return_alias.clone()))
+                    .equals(object_dependency::Column::Oid),
+            )
+            .to_owned();
+
+        let common_table_expr = CommonTableExpression::new()
+            .query(
+                base_query
+                    .clone()
+                    .union(UnionType::All, cte_referencing)
+                    .to_owned(),
+            )
+            .column(cte_return_alias.clone())
+            .table_name(cte_alias.clone())
+            .to_owned();
+
+        let query = SelectStatement::new()
+            .distinct()
+            .column(cte_return_alias.clone())
+            .from(cte_alias)
+            .order_by(cte_return_alias.clone(), Order::Desc)
+            .to_owned()
+            .with(
+                WithClause::new()
+                    .recursive(true)
+                    .cte(common_table_expr)
+                    .to_owned(),
+            )
+            .to_owned();
+
+        let (sql, values) = query.build_any(&*inner.db.get_database_backend().get_query_builder());
+        let res = inner
+            .db
+            .query_all(Statement::from_sql_and_values(
+                inner.db.get_database_backend(),
+                sql,
+                values,
+            ))
+            .await?;
+
+        let ids: Vec<i32> = res
+            .into_iter()
+            .map(|row| row.try_get("", &cte_return_alias.to_string()).unwrap())
+            .collect_vec();
+        Ok(ids)
+    }
+
+    pub async fn drop_database(
+        &self,
+        database_id: DatabaseId,
+    ) -> MetaResult<(ReleaseContext, NotificationVersion)> {
+        let inner = self.inner.write().await;
```

Comment:
This is interesting. Since the database can handle concurrency for us, do we still need an exclusive access to the controller?

Reply:
I tried to avoid an exclusive access to the controller, but consistency of the notification service still need it. This's a quick start to make it run first. I think we can do some refactoring on the notification side to avoid it in the future.

Reply:
I think we shouldn't completely rely on the concurrency control of databases. Basically, the database is only responsible for its data, while RisingWave (as an application) has more and higher limitations.